### PR TITLE
nixos/tests: nginx-http3 test

### PR DESCRIPTION
**Pull request by Izorkin**
###### Description of changes
Add simple test curl package.
Included tests with http3 protocol.

cc @ajs124

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Izorkin** [comment]
@GrahamcOfBorg test curl

**Comment by ajs124** [comment]
Hm, do we maybe want to integrate this with some other test, e.g. an nginx http3 test, because it tests that as much as it tests curl.

**Comment by Izorkin** [comment]
Split into 2 tests - curl and nginx-http3?

**Comment by Izorkin** [comment]
@GrahamcOfBorg test curl nginx-http3

**Comment by Mindavi** [comment]
Tests run fine for me:

nix build github:Izorkin/nixpkgs/init-tests-curl#nixosTests.nginx-http3 -L
nix build github:Izorkin/nixpkgs/init-tests-curl#nixosTests.curl -L

**Comment by Izorkin** [comment]
Added header read check.

**Comment by Izorkin** [comment]
cc @ajs124 @mweinelt @risicle

**Review by Mindavi — APPROVED** [review]
At least they run fine and seems useful to me to add these kind of smoke tests to get a grip on how well the packages are running.

**Comment by risicle** [comment]
Total coincidence  #170828

**Comment by Izorkin** [comment]
@risicle leave only `nginx-http3`?

**Comment by risicle** [comment]
If we do, should probably still reference it from `curl.passthru.tests` as it does test some curl functionality that likely doesn't get heavily tested otherwise.

**Comment by Izorkin** [comment]
This variant?
```
  passthru = {
    tests.nginx-http3 = nixosTests.nginx-http3;
    inherit opensslSupport openssl;
  };
```

**Comment by risicle** [comment]
Yes I think that makes sense.

**Comment by Izorkin** [comment]
@risicle update PR.

**Comment by Mindavi** [comment]
The test now fails for me; nix build github:izorkin/nixpkgs/init-tests-curl#nixosTests.nginx-http3

**Comment by Izorkin** [comment]
@Mindavi works for me:
```
nix-build nixos/tests/nginx-http3.nix
...
client: must succeed: curl --verbose --http3 --user-agent 'Curl test 3.0' https://site.local
client #   % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
client #                                  Dload  Upload   Total   Spent    Left  Speed
client #   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0*   Trying 192.168.2.102:443...
client # * Connect socket 5 over QUIC to 192.168.2.102:443
client # *  CAfile: /etc/ssl/certs/ca-certificates.crt
client # *  CApath: none
client #   0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0*  subjectAltName: host "site.local" matched cert's "site.local"
client # * Verified certificate just fine
client # * Connected to site.local () port 443 (#0)
client # * h2h3 [:method: GET]
client # * h2h3 [:path: /]
client # * h2h3 [:scheme: https]
client # * h2h3 [:authority: site.local]
client # * h2h3 [user-agent: Curl test 3.0]
client # * h2h3 [accept: */*]
client # * Using HTTP/3 Stream ID: 0 (easy handle 0x234c840)
client # > GET / HTTP/3
client # > Host: site.local
client # > user-agent: Curl test 3.0
client # > accept: */*
client # >
client # < HTTP/3 200
client # < server: nginx
client # < date: Sun, 01 May 2022 20:05:12 GMT
client # < content-type: text/html
client # < content-length: 39
client # < etag: "4661zjkxbbr39lnzwzgyh7j46gcqs73a"
client # < alt-svc: h3=":443"; ma=86400
client # < accept-ranges: bytes
client # <
client # { [39 bytes data]
client # 100    39  100    39    0     0     87      0 --:--:-- --:--:-- --:--:--    87
client # * Connection #0 to host site.local left intact
(finished: must succeed: curl --verbose --http3 --user-agent 'Curl test 3.0' https://site.local, in 0.51 seconds)
server: must succeed: cat /var/log/nginx/access.log | grep 'Curl test 3.0'

...
```

**Comment by Mindavi** [comment]
Maybe related to some server I'm running? I can reliably reproduce the build error.

**Comment by Mindavi** [comment]
@ofborg test curl

**Comment by Izorkin** [comment]
@GrahamcOfBorg test nginx-http3

**Comment by Izorkin** [comment]
Removed the curl test, left only the test with the http3 protocol.

**Comment by Izorkin** [comment]
Test fixed in PR #172576

**Comment by Mindavi** [comment]
Oh interesting! So it was actually broken?

File: nixos/tests/nginx-http3.nix [hunk 1]
```diff
@@ -0,0 +1,145 @@
+import ./make-test-python.nix ({lib, pkgs, ...}:
+let
+  test-certificates = pkgs.runCommandLocal "test-certificates" { } ''
```
Comment (lheckemann) [inline nixos/tests/nginx-http3.nix:3]: We should hardcode certificates here, for the reasons explained in https://github.com/NixOS/nixpkgs/pull/91121#discussion_r505410235

Comment (Izorkin) [reply nixos/tests/nginx-http3.nix:3]: I did not find a way to execute command `step` in`files` block.

**Comment by Izorkin** [comment]
@GrahamcOfBorg test nginx-http3

File: nixos/tests/nginx-http3.nix (hunk 1 above)
Comment (Izorkin) [reply nixos/tests/nginx-http3.nix]: Fixed!

**Comment by Izorkin** [comment]
@GrahamcOfBorg test nginx-http3

**Comment by Izorkin** [comment]
@GrahamcOfBorg test nginx-http3

**Comment by Izorkin** [comment]
@GrahamcOfBorg test nginx-http3

**Comment by Izorkin** [comment]
@GrahamcOfBorg test nginx-http3

**Comment by Izorkin** [comment]
@GrahamcOfBorg test nginx-http3

**Pull request merged**